Improve docs for custom printing

rasto2211 opened the pull request:
Add example of writing custom printing method for printing custom expressions.

Simplify some parts of the example to highlight the important parts.

Rewrite the code from python2 to python3.

Encourage the users to write their own wrappers for `print()` instead of
overriding `Basic.__str__` .

Improve the formatting of the whole tutorial.

Move the parts that were duplicated into one place.

Fixes #11025

rasto2211 committed:
Improve docs for custom printing

Add example of writing custom printing method for printing custom expressions.

Simplify some parts of the example to highlight the important parts.

Rewrite the code from python2 to python3.

Encourage the users to write their own wrappers for `print()` instead of
overriding `Basic.__str__` .

Improve the formatting of the whole tutorial.

Move the parts that were duplicated into one place.

Fixes #11025

rasto2211 commented:
It's a bit weird that the job #22335.29 has failed. It's the only one failing. The tail of the log is:
```
sympy/matrices/tests/test_matrices.py[2] ..                                 [OK]
sympy/polys/tests/test_polyroots.py[1] .                                    [OK]
sympy/simplify/tests/test_hyperexpand.py[16] ..
The job exceeded the maximum time limit for jobs, and has been terminated.
```
When I rerun `test_hyperexpand.py` on my machine using `bin/test simplify/` then I see that the test is passing. Isn't there the problem that the deadline for the test is too strict? Besides that I don't think I affected the test by my changes. I just changed the documentation. Could please anyone explain me what's the problem?

jksuom commented:
> It's a bit weird that the job #22335.29 has failed.

That is a Travis issue. It seems that there are too many tests in that split.

rasto2211 commented:
Is there already an issue on this problem or should I file a new issue?

asmeurer commented on sympy/printing/printer.py:
```diff
@@ -55,135 +61,115 @@
-    1. Subclass Printer.
+        def _latex(self, printer=None):
+            # This is the printmethod of LatexPrinter.
+            return \"\\\\textbf{{{}}}\".format(self.text)
```
Better to show using a raw string in the latex example.

rasto2211 replied:
Fixed.

asmeurer commented on sympy/printing/printer.py:
```diff
@@ -55,135 +61,115 @@
-    2. Define Printer.printmethod in your subclass.
-       If a object has a method with that name, this method will be used
-       for printing.
+    class MyPrinter(Printer):
```
We might want to separate these, as you generally want to do one or the other.

asmeurer replied:
Usually you want to subclass an existing printer, like StrPrinter or LatexPrinter.

rasto2211 replied:
Well, I think that it is nice to have custom printing method and custom printer in one example because this way you can see the interaction of custom expression and custom printer. You can see that `MyBasic` does not have the method `_myprinter` so it cannot call it and instead the printing is dispatched by `_print_MyBasic`. 

The original example had `_print_MyClass(self, expr)` but `MyClass` wasn't defined anywhere so you could not compile the example. Therefore I also decided to add the code for `MyClass` and I renamed `MyClass` to `MyBasic` because it subclasses `sympy.core.basic`. 

By "Usually you want to subclass an existing printer, like StrPrinter or LatexPrinter." you mean that we also want to have `StrPrinter` in the example instead of `Printer` or you just want me to add a comment on this thing. I don't see a reason why directly subclassing `Printer` is a bad idea.

asmeurer replied:
> The original example had _print_MyClass(self, expr) but MyClass wasn't defined anywhere so you could not compile the example. Therefore I also decided to add the code for MyClass and I renamed MyClass to MyBasic because it subclasses sympy.core.basic.

I think the subclassed printer should override an existing SymPy function. It doesn't really make sense to subclass a printer for a custom class, because you can just define the printer on the class. So in this example, it is a little unclear why you would even want to subclass a printer, and why we have both ways. But generally the recommendations are:

- If you have a custom SymPy class, add the printmethod to that class. 
- If you want to change the printing for an existing SymPy class, subclass the appropriate printer(s).

> By "Usually you want to subclass an existing printer, like StrPrinter or LatexPrinter." you mean that we also want to have StrPrinter in the example instead of Printer or you just want me to add a comment on this thing. I don't see a reason why directly subclassing Printer is a bad idea.

Again, it's not a realistic example. The real use-case is to change how the existing printers work. StrPrinter and LatexPrinter are the most common (also the pretty printer, but that also has some advanced stuff because it use to use the pretty form stuff). 

You can see lots of examples of things people want to do in [StackOverflow](https://stackoverflow.com/search?q=%5Bsympy%5D+printer).  [Here's](https://stackoverflow.com/questions/43350381/sympy-modifying-latex-output-of-derivatives/43356409#43356409) an example of the sort of thing that could make a nice example, becaues it's short, and it's also something that people actually would want to do.

asmeurer replied:
By the way the pretty printer stuff itself should be documented somewhere, but that can be done separately from this PR.

asmeurer commented on sympy/printing/printer.py:
```diff
@@ -55,135 +61,115 @@
+        def _print_MyBasic(self, expr):
+            # Since MyBasic does not implement _myprinter method, this method
+            # will be called instead.
+            return str(expr).upper()
```
Printers should generally never call `str`. They should use the printer itself to recursively print the object, with `self._print`. This should also be noted in the text above if it isn't already.

rasto2211 replied:
Well, I think that in this example we really want to call `str()` on `MyBasic`. We basically need to get the `expr.text` and I think that it is better to write `str(expr)` than `expr.text`. `MyBasic` is not composite expression. It just holds a string.

Hmmm... Maybe I could add an example of using recursive `Printer._print` on a composite expression. What do you think? Wouldn't that just make the example too complicated?

The original documentation was also using `str()` in `_print_Derivative()`

asmeurer replied:
I think if we change it to my suggestion [above](https://stackoverflow.com/questions/43350381/sympy-modifying-latex-output-of-derivatives/43356409#43356409), then it is more natural. The _print_Derivative example it just complicated enough to show that, but simple enough that it doesn't have a lot of code to distract.

moorepants replied:
I agree, we should use `self._print()` in this documentation to influence good practice. `str()` is bound to break on many expressions that are complicated.

asmeurer replied:
The point is that str won't call the methods for the custom printer itself. Consider a nested Derivative for example.

rasto2211 committed:
Show raw string in latex example

rasto2211 committed:
Use more realistic example for custom printer

https://stackoverflow.com/questions/43350381/sympy-modifying-latex-output-of-derivatives/43356409#43356409

rasto2211 commented:
I've added more realistic example of a custom printer from [here](https://stackoverflow.com/questions/43350381/sympy-modifying-latex-output-of-derivatives/43356409#43356409).

Unfortunately, I did not find any realistic example of a custom expression when we need to write our own printing method. Any ideas?

asmeurer commented:
Let me think about it. What was the example that was there before?

rasto2211 commented:
Originally, there was an example with a string expression. You can see it [here](http://docs.sympy.org/latest/modules/printing.html)

asmeurer commented:
The simplest thing would be some function subclass. If you can't think of something you could just use something that already exists, just for example purposes, like say, factorial.

rasto2211 committed:
Use Mod as an example for a custom printing method

rasto2211 commented:
@asmeurer I finally chose to use modulo operation. I think that it is quite realistic example. You might want to print modulo in latex in a different way than it is implemented in SymPy.

smichr commented on sympy/printing/printer.py:
```diff
@@ -55,135 +61,111 @@
-from functools import cmp_to_key
+.. _printer_example:
 
+In the example below, we have a printer which prints derivative of a function
```
-> 
In the example below, we have a printer which prints the derivative of a function
in a shorter form.

(added "the" and dropped redundant reference to "code below".)

smichr commented on sympy/printing/printer.py:
```diff
@@ -55,135 +61,111 @@
-    3. In your subclass, define ``_print_<CLASS>`` methods
+    def print_my_latex(expr):
+        \"\"\" Most of the printers define their own wrappers for print().
+        These wrappers usually takes printer settings. Our printer does not have
```
takes -> take

smichr commented on sympy/printing/printer.py:
```diff
@@ -55,135 +61,111 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
-    4. If convenient, override ``self.emptyPrinter``
+In this example below, we want to print modulo operator in a different way in
```
->
In the example below, the latex printing of the modulo operator is modified. This
is done by overriding the method ``_latex`` of ``Mod``.

smichr commented on sympy/printing/printer.py:
```diff
@@ -4,14 +4,26 @@
+
+1. **Let the object print itself if it knows how.**
+
+    The printers looks for a specific method in every object. The name of that method
```
looks -> look

smichr commented:
Just a few recommended changes and then I think this is ready to go.

asmeurer commented on sympy/printing/printer.py:
```diff
@@ -122,6 +122,32 @@ def print_my_latex(expr):
+    class ModOp(Mod):
+        def _latex(self, printer=None):
+            a, b = [printer.doprint(i) for i in self.args]
+            return "\operatorname{Mod}{\left( %s,%s \\\\right)}" % (a,b)
```
This needs to be 

```
return r"\\operatorname{Mod}{\\left( %s,%s \\\\right)}" % (a,b)
```

The `r` should always be used for latex strings. The extra backslashes are needed because the docstring itself is not a raw string (alternately, you could convert the docstring to a raw string). 

I would build the docs and double check that these latex examples render with the correct number of backslashes. Also, the next time you commit, the Travis tests will fail on this because we are now testing against invalid escapes like `\o`.

asmeurer commented:
It would be useful to point out in each example the importance of recursively printing with `self._print`/`printer.doprint`. Maybe an example could show a nested Derivative/Mod and demonstrate that if you just naively use `latex` to recursively print (or `str` as would be the case for a string printer), then the custom printer itself won't apply. 

This is the number one gotcha I see: people using `str` (or the equivalent `%s`) in a custom str printer instead of the printer itself to recursively print.

rasto2211 committed:
Fix some typos and remove redundant sentences

rasto2211 commented:
Fixed the problems that @smichr brought up. I'll fix the problems that @asmeurer mentioned in the next commit.

rasto2211 committed:
Fix problem with escaping latex string

gxyd commented:
@rasto2211 just a polite ping, good work till now. There are still a few valid failing tests here.

rasto2211 committed:
Add example for nested modulo

rasto2211 committed:
Add comment about nested expression to the custom printer example

rasto2211 commented:
I added an example of nested modulo but we cannot really have an example of nested derivative since it does not make sense with the shortened notation that we introduced in the example. I just added a comment about nested expressions into the code to emphasize the fact that it's better to use `self._print` than `str()` or `latex()`.

rasto2211 committed:
Remove unnecessary new line

rasto2211 committed:
Fix addional problems with escaping special characters

rasto2211 committed:
Replace tabs with spaces

rasto2211 commented:
I have no idea what I've done wrong right now so the travis job is failing. It fails with this error:
```
The command "curl -s -o python-3.6.tar.bz2 ${archive_url}" failed and exited with 56 during .
```

ylemkimon commented:
It is a Travis error that it couldn’t install Python 3.6 and isn’t related to the PR. Restarted the job.

gxyd commented on sympy/printing/printer.py:
```diff
@@ -4,14 +4,26 @@
+
+1. **Let the object print itself if it knows how.**
+
+    The printers look for a specific method in every object. The name of that method
```
"The printers look" -> "The printer looks"?

rasto2211 replied:
Fixed.

gxyd commented on sympy/printing/printer.py:
```diff
@@ -55,135 +61,131 @@
+    Its job is to provide infrastructure for implementing new printers easily.
 
+    If you want to define your custom Printer or your custom printing method
+    for your custom class then see the example above: printer_example_ .
```
an underscore is need here?  `printer_example_` shouldn't it be `_printer_example`?

rasto2211 replied:
Well, that's the way internal links in sphinx work. When I changed it to `_printer_example` then the link stopped working.

gxyd commented:
Otherwise LGTM.

rasto2211 committed:
Change singular to plural

gxyd commented:
I'll commit this in 48 hours, unless anyone has any comments.

gxyd commented:
Thanks @rasto2211 for your contribution. Good work :sparkles:
